ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,192 @@
+    }
+
+    @Override
+    public void onDeregister(final QueryMetadata query) {
+        final KafkaStreams streams = query.getKafkaStreams();
+        kafkaStreams.remove(streams);
+        for (ThreadMetadata thread : streams.localThreadsMetadata()) {
+            final String name = thread.threadName();
+            previousFlushTime.remove(name);
+            previousConsumerIOTime.remove(name);
+            previousProducerBufferWaitTime.remove(name);
+        }
+    }
+
+    @Override
+    public void run() {
+        logger.info("Reporting CSU thread level metrics");
+        processingRatio();
+
+    }
+
+    // public for testing
+    public double processingRatio() {
+        long sampleTime = time.milliseconds();
+
+        final long windowSize = sampleTime - lastSampleTime;
+        final long windowStart = lastSampleTime;
+        double blockedTime;
+        blockedTime = sampleTime - lastSampleTime;
+        for (KafkaStreams stream : kafkaStreams) {
+            for (ThreadMetadata thread : stream.localThreadsMetadata()) {
+                final double threadTimes = getProcessingRatio(thread.threadName(), stream, windowStart, windowSize);
+                blockedTime = Math.min(threadTimes, blockedTime);
+            }
+        }
+        final double notBlocked = windowSize - blockedTime;
+        lastSampleTime = sampleTime;
+        logger.info("The current processing ratio is " + percentage(notBlocked, windowSize) + "%");
+        return percentage(notBlocked, windowSize);
+    }
+
+    // public for testing
+    public double getProcessingRatio(final String threadName, final KafkaStreams streams, final long windowStart, final double windowSize) {
+        try {
+            final Collection<? extends Metric> allMetrics = streams.metrics().values();
+            final Map<String, Double> consumerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(CONSUMER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && !m.metricName().tags().getOrDefault("client-id", "").contains("restore-consumer")
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> restoreConsumerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(CONSUMER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && m.metricName().tags().getOrDefault("client-id", "").contains("restore-consumer")
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> producerMetrics = allMetrics.stream()
```

Comment:
I think we should consider admin client's blocking portion as well?

Reply:
What does the streams thread use the admin client for during regular processing?

Reply:
During rebalances, we use admin clients to get end offsets, and topic metadata, which are all blocking calls.

Reply:
Ah makes sense - I'd think we should be able to use the same network client metrics from the admin client here.

Reply:
To make sure I'm following - gather the same metric names but also from the admin client? So that would be a different metric group, right?